### ISSUE TITLE
flasher boot: invalid boot image

### ISSUE THREAD

**Issue by joshas** [postmarketOS/pmbootstrap#1592]
Finally got kernel image compiled for Huawei U8160 (Vodafone 858 Smart), but now I'm stuck on booting kernel with fastboot. After running: `$ ./pmbootstrap.py flasher boot` I got following error message:
````
[22:32:15] (rootfs_huawei-u8160) mkinitfs huawei-u8160
[22:32:18] (native) boot huawei-u8160 kernel
downloading 'boot.img'...
OKAY [  1.020s]
booting...
FAILED (remote: invalid boot image)
finished. total time: 1.026s
[22:32:20] ERROR: Command failed: (native) % fastboot -i  -c mem=211M console=ttyMSM2,115200n8 androidboot.hardware=huawei console=ttyUSBCONSOLE0 androidboot.console=ttyUSBCONSOLE0 boot /mnt/rootfs_huawei-u8160/boot/boot.img-huawei-u8160
Run 'pmbootstrap log' for details.
````
Same error when using full disk encryption, or disabling it. Setting vendor_id in deviceinfo makes fastboot stop with `<waiting for device>` message.
Device is quite old and only has 2.6.32.63 kernel available. Any leads on how to debug this non informative error message would be helpful.

I can successfully boot original boot.img and recovery.img using fastboot boot command.

Tried booting with original kernel image and pmOS initramfs - invalid boot image error. Could it be, that pmboostrap produces incompatible initramfs image? While I can unpack pmOS initramfs image, I cannot do that with original initramfs image, file command describes its contents as "data".

**Comment by ollieparanoid** [comment]
Welcome to postmarketOS, @joshas!

Some ideas:
* Have you generated the fastboot offsets with pmbootstrap? You can double-check them with `pmbootstrap bootimg_analyze path/to/boot.img`
* Could it be, that your phone only accepts signed boot.img files?
  * If you can find custom ROMs or recoveries (TWRP/Cyanogen MOD/... ) that work with your device, then that's unlikely.
  * Even if that is the case, there may still be some tricks we can pull (e.g. boot another kernel from the signed one)
* Maybe a special boot.img header is required, [like for the fairphone 1](https://github.com/postmarketOS/pmbootstrap/pull/1566#discussion_r195895488).
  * Consider taking a look at the original `boot.img` with a hex editor - are there lots of `00` somewhere at the beginning?
* Is there a big difference in file size of the original and the pmOS boot.img file?
* Try extracting both boot.img files and look for differences (e.g. with `unpackbootimg` as packaged in `mkbootimg-osm0sis` in pmOS)
* It might be that `pmbootstrap flasher flash_kernel` works (as everything, on your own risk)

Found your error message on these sites:
* https://forum.xda-developers.com/showthread.php?t=819104 (4k start block, like the FP1)
* https://oliverkuster.wordpress.com/2011/03/18/motorola-xt300spice-juicy-details/ (signed)

**Comment by joshas** [comment]
Thanks for ideas, they helped me to narrow problem to ramdisk (initramfs) image size. Here's a quick comparison of boot.img contents:

File | original boot.img | custom ROM | postmarketOS
----|-----|----|-----
ramdisk.cpio.gz | 270 KB | 163 KB | **1.1 MB**
zImage (kernel) | 2.5 MB | 3.0 MB | 2.9 MB

As I copied offset values from original boot.img, I think it expects to find files in similar sizes to original. And I'm not sure how big boot.img partition is on device, postmarketOS boot.img might just not fit there at all.
Is there any way to build smaller postmarketOS ramdisk? I'll try to figure out how big boot partition is on device and if changing fastboot offsets might fix booting issue.

**Comment by ollieparanoid** [comment]
Seems like the size of the initramfs/boot.img is not the problem after all, but here is in theory how to make it smaller:
* extract the initramfs that has been generated and look for big files (e.g. `pmbootstrap initfs extract` and using `ncdu`)
* adjust [`mkinitfs.sh`](https://github.com/postmarketOS/pmbootstrap/blob/master/aports/main/postmarketos-mkinitfs/mkinitfs.sh) to throw out what you don't need and what is big in size
* compile a minimal busybox version to save some more space
* try more efficient compression algorithms for the kernel, it will make the overall `boot.img` file size smaller as well.

> Another thing I tried: using abootimg update one of booting boot images (original from device and one from custom ROM) with pmOS built kernel and initramfs. To my surprise fastboot boot accepts updated image when using custom ROM boot.img, but does not boot.

That's a good method for debugging! Here are some tips:
https://wiki.postmarketos.org/wiki/Troubleshooting:boot#Kernel_doesn.27t_even_boot

**Comment by joshas** [comment]
Looks like I have the same issue as #1203 - MSM7225 SoC doesn't have FPU. Should've done more research first. But at least I learned quite a lot postmarketOS build process and debugging tools.
Would be nice to put not supported devices in separate table, because only now I noticed htc-buzz in "Not booting" devices list.
@ollieparanoid thanks for your help, without it I'd be still stuck without any idea why this doesn't work.
